src/gateway/routers/RouterAggregator.sol

```diff
@@ -0,0 +1,231 @@
+    function handle(bytes calldata payload) public {
+        Router memory router = validRouters[msg.sender];
+        require(router.id != 0, "RouterAggregator/invalid-router");
+        _handle(payload, router);
+    }
+
+    function _handle(bytes calldata payload, Router memory router) internal {
+        if (MessagesLib.isRecoveryMessage(payload)) {
+            require(routers.length > 1, "RouterAggregator/no-recovery-with-one-router-allowed");
+            return _handleRecovery(payload);
+        }
+
+        if (router.quorum == 1 && !MessagesLib.isMessageProof(payload)) {
+            // Special case for gas efficiency
+            gateway.handle(payload);
+            emit ExecuteMessage(payload, msg.sender);
+            return;
+        }
+
+        bytes32 messageHash;
+        ConfirmationState storage state;
+        if (MessagesLib.isMessageProof(payload)) {
+            messageHash = MessagesLib.parseMessageProof(payload);
+            state = _confirmations[messageHash];
+            state.proofs[router.id - 1]++;
+
+            emit HandleProof(messageHash, msg.sender);
+        } else {
+            messageHash = keccak256(payload);
+            state = _confirmations[messageHash];
+            state.messages[router.id - 1]++;
+
+            emit HandleMessage(payload, msg.sender);
+        }
+
+        if (state.messages.countNonZeroValues() >= 1 && state.proofs.countNonZeroValues() >= router.quorum - 1) {
+            // Reduce total message confiration count by 1, by finding the first non-zero value
+            state.messages.decreaseFirstNValues(1, 1);
+
+            // Reduce total proof confiration count by quorum
+            state.proofs.decreaseFirstNValues(router.quorum, 1);
+
+            if (MessagesLib.isMessageProof(payload)) {
+                gateway.handle(pendingMessages[messageHash]);
+
+                // Only if there are no more pending messages, remove the pending message
+                if (state.messages.isEmpty() && state.proofs.isEmpty()) {
+                    delete pendingMessages[messageHash];
+                }
+            } else {
+                gateway.handle(payload);
+            }
+
+            emit ExecuteMessage(payload, msg.sender);
+        } else if (!MessagesLib.isMessageProof(payload)) {
+            pendingMessages[messageHash] = payload;
+        }
+    }
+
+    /// @dev Governance on Centrifuge Chain can initiate message recovery. After the challenge period,
+    ///      the recovery can be executed. If a malign router initiates message recovery, governance on
+    ///      Centrifuge Chain can dispute and immediately cancel the recovery, using any other valid router.
+    ///
+    ///      Only 1 recovery can be outstanding per message hash. If multiple routers fail at the same time,
+    //       these will need to be recovered serially (increasing the challenge period for each failed router).
```

Comment:
In the scenario that 1 router stops working (is not forwarding messages any messages anymore). The router would be replaced (using ScheduleRely most likely), but to do so you'd need to use the other bridges as recovery

Reply:
I see, so it's a timelocked way of bypassing quorom, in case quorum is impossible due to a router malfunction? In which case, theres still a denial of service risk, but I guess it's a risk we have to accept.